### DISCUSSION TITLE
Job Skipped Unnecessarily

### DISCUSSION THREAD

**Discussion by andynicklen** [community/community#25224] [Actions]
<p>Hi, I have a job <strong>job_test</strong> which is being skipped for some reason. It has a needs dependency on the job <strong>pre_install</strong>, which is set to always run. I cant work out why it’s not running.</p>
<p>I’ve confirmed that the output from pre_install is being output, but this shouldnt prevent <strong>job_test</strong>  from running. The other jobs further up are being legitimately skipped, so i’m not concerned about these ones. Thoughts much appreciated.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/3/2/32eab547f36aa2752a0bce3c713fe1a727dc9a55.png" data-download-href="/uploads/short-url/7gqJt4ARGk3ooaQHjqMDV3BbSWp.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/28435773/181089680-f071f98a-beb5-43f5-bf29-9e9bfd885c72.png" alt="image" data-base62-sha1="7gqJt4ARGk3ooaQHjqMDV3BbSWp" width="596" height="499" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/3/2/32eab547f36aa2752a0bce3c713fe1a727dc9a55_2_596x499.png, https://aws1.discourse-cdn.com/github/original/3X/3/2/32eab547f36aa2752a0bce3c713fe1a727dc9a55.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/3/2/32eab547f36aa2752a0bce3c713fe1a727dc9a55.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/2/32eab547f36aa2752a0bce3c713fe1a727dc9a55_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">796×667 60.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p><img src="https://user-images.githubusercontent.com/28435773/181089681-683648e9-e15c-47c4-bec1-ffe38ed800eb.png" alt="image" data-base62-sha1="gUcdYeV5MTtMUesbEjTtzzRh19z" width="248" height="347"></p>

**Comment by andynicklen** [comment]
<p>I found a solution to this <a href="https://github.community/t/jobs-being-skipped-while-using-both-needs-and-if-unclear-why/18480">here</a>. I’d describe as a workaround and i’d hope Github havent designed it like this on purpose.</p>
<p>In my example (I assume) <strong>job_test</strong> isnt executing because dependencies further up have been skipped, even though <strong>job_test</strong> itself is only dependent on <strong>pre_install</strong>, which ran successfully.</p>
<p>The workaround for job_test was to add:<br>
<code> if: always() &amp;&amp; needs.pre_install.result == 'skipped'</code></p>
<p>To me this reads - always run job_test but also evaluate the condition to the right. So in effect, it makes it conditional. This seems like a crazy way of acheiving something that could be very simple and far less confusing :confused:</p>

**Comment by andynicklen** [comment, marked as answer]
<p>Github support shed some light on this and will updating their documentation so the behaviour is clearer.</p>
<p>Here is a single job…</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/a/7/a7d05e79178276067625791b3f87c772a37857da.png" data-download-href="/uploads/short-url/nWyaiJxzZbQrjpktqrPbcItWl6i.png?dl=1" title="Screenshot at Apr 05 13-08-05" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/28435773/181089683-8f655143-c774-41ac-9da0-090a0b32b46f.png" alt="Screenshot at Apr 05 13-08-05" data-base62-sha1="nWyaiJxzZbQrjpktqrPbcItWl6i" width="376" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/a/7/a7d05e79178276067625791b3f87c772a37857da_2_376x500.png, https://aws1.discourse-cdn.com/github/optimized/3X/a/7/a7d05e79178276067625791b3f87c772a37857da_2_564x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/a/7/a7d05e79178276067625791b3f87c772a37857da.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/a/7/a7d05e79178276067625791b3f87c772a37857da_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Screenshot at Apr 05 13-08-05</span><span class="informations">670×890 201 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>If a dependent step with an if: condition is <em>skipped</em>, subsequent steps are <strong>not</strong> skipped. You can see this above.</p>
<p>However, chaining of jobs behaves differently. If you imagine the above scenario as jobs instead of steps, the <strong>checkout repo npm install job</strong> would never run (unless you force it to run using if: always). In effect, the workflow becomes cancelled once a job is skipped. Github say this is to speed up performance.</p>
<p>My solution now consist of two jobs and all of the if: logic is within the steps.</p>
<p>They key takeaway is to remember that if any of your <em>jobs</em> are likely to be skipped, any subsequent jobs will not run.</p>
<p>There are workarounds including use of the <a href="https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run" rel="noopener nofollow ugc">workflow_run</a> event, or even creating your own <a href="https://docs.github.com/en/rest/reference/actions#create-a-workflow-dispatch-event" rel="noopener nofollow ugc">workflow_dispatch</a> events. Both of these seemed overkill for what I needed.</p>
<p>It’s worth pointing out - at the time of writing the use of if: always() will mean your jobs continue to run even if you press the cancel/stop button in the console :thinking:</p>

**Reply by brianjmurrell** [reply]
> It’s worth pointing out - at the time of writing the use of if: always() will mean your jobs continue to run even if you press the cancel/stop button in the console 🤔

Which makes the use of `if: always()` a complete non-starter.  Users (or their admins) have to be able to cancel their jobs.  But one still needs to be able to run subsequent jobs even if previous jobs have been skipped.

**Reply by James-Mart** [reply]
I was able to use `if: ${{ !cancelled() && ...` which still allows me to cancel.
